src/codegen/ast.lisp

```diff
@@ -428,97 +430,111 @@ both CL namespaces appearing in NODE"
-   :rator (node-direct-application-rator node)
-   :rands (mapcar
-           (lambda (node)
-             (tc:apply-substitution subs node))
-           (node-direct-application-rands node))))
-
-(defmethod tc:apply-substitution (subs (node node-abstraction))
-  (make-node-abstraction
-   :type (tc:apply-substitution subs (node-type node))
-   :vars (node-abstraction-vars node)
-   :subexpr (tc:apply-substitution subs (node-abstraction-subexpr node))))
-
-(defmethod tc:apply-substitution (subs (node node-let))
-  (make-node-let
-   :type (tc:apply-substitution subs (node-type node))
-   :bindings (loop :for (name . node) :in (node-let-bindings node)
-                   :collect (cons name (tc:apply-substitution subs node)))
-   :subexpr (tc:apply-substitution subs (node-let-subexpr node))))
-
-(defmethod tc:apply-substitution (subs (node node-lisp))
-  (make-node-lisp
-   :type (tc:apply-substitution subs (node-type node))
-   :vars (node-lisp-vars node)
-   :form (node-lisp-form node)))
-
-(defmethod tc:apply-substitution (subs (node match-branch))
-  (make-match-branch
-   :pattern (match-branch-pattern node)
-   :body (tc:apply-substitution subs (match-branch-body node))))
-
-(defmethod tc:apply-substitution (subs (node node-match))
-  (make-node-match
-   :type (tc:apply-substitution subs (node-type node))
-   :expr (tc:apply-substitution subs (node-match-expr node))
-   :branches (mapcar
-              (lambda (node)
-                (tc:apply-substitution subs node))
-              (node-match-branches node))))
-
-(defmethod tc:apply-substitution (subs (node node-seq))
-  (make-node-seq
-   :type (tc:apply-substitution subs (node-type node))
-   :nodes (mapcar
-           (lambda (node)
-             (tc:apply-substitution subs node))
-           (node-seq-nodes node))))
-
-(defmethod tc:apply-substitution (subs (node node-return))
-  (make-node-return
-   :type (tc:apply-substitution subs (node-type node))
-   :expr (tc:apply-substitution subs (node-return-expr node))))
-
-(defmethod tc:apply-substitution (subs (node node-field))
-  (make-node-field
-   :type (tc:apply-substitution subs (node-type node))
-   :name (node-field-name node)
-   :dict (tc:apply-substitution subs (node-field-dict node))))
-
-(defmethod tc:apply-substitution (subs (node node-dynamic-extent))
-  (make-node-dynamic-extent
-   :type (tc:apply-substitution subs (node-type node))
-   :name (node-dynamic-extent-name node)
-   :node (tc:apply-substitution subs (node-dynamic-extent-node node))
-   :body (tc:apply-substitution subs (node-dynamic-extent-body node))))
-
-(defmethod tc:apply-substitution (subs (node node-bind))
-  (make-node-bind
-   :type (tc:apply-substitution subs (node-type node))
-   :name (node-bind-name node)
-   :expr (tc:apply-substitution subs (node-bind-expr node))
-   :body (tc:apply-substitution subs (node-bind-body node))))
+(defgeneric copy-node (node &key)
```

Comment:
I think this will have subtle bugs if someone passes in nil. You'll need to do the laborious thing of using suppliedp arguments...

Reply:
yeah, I checked that even `node-literal-value` can't be a boolean, so I think all the cases should be fine, but the type updating is all i _really_ wanted, so I went ahead and reverted and stashed this generic, going back to the `copy-node (node &optional type)` we had previously